jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -5466,12 +5494,255 @@ public Object helper(Object[] arguments) throws Throwable {
+		MethodType targetType = target.type();
+		MethodType filterType = filter.type();
+		MethodType resultType = targetType.changeReturnType(filterType.returnType());
+		int filterParamCount = filterType.parameterCount();
+		if (filterParamCount > 1) {
+			for (int i = 0; i < filterParamCount - 1; i++) {
+				resultType = resultType.appendParameterTypes(filterType.parameterType(i));
+			}
+		}
+		return buildTransformHandle(new CollectReturnHelper(target, filter), resultType);
+	}
+	
+	/**
+	 * Return an array of checked exceptions for a MethodHandle.
+	 * 
+	 * @param mh A MethodHandle to lookup for checked exceptions.
+	 * 
+	 * @return an array of checked exception classes if any or null.
+	 * 
+	 * @throws InternalError for an error.
+	 */
+	static Class<?>[] checkedExceptions(MethodHandle mh) {
+		if (mh == null) {
+			return null;
+		}
+		List<MethodHandle> relatedMHsList = null;
+		List<Class<?>> exceptionTypesList = null;
+		while (mh != null) {
+			if (mh instanceof PrimitiveHandle) {
+				Class<?>[] exceptionTypes = null;
+				MethodHandleInfoImpl mhi = new MethodHandleInfoImpl((PrimitiveHandle)mh);
+				if (mhi.isConstructor()) {
+					exceptionTypes = mhi.reflectAs(Constructor.class, Lookup.IMPL_LOOKUP).getExceptionTypes();
+				} else if (mhi.isMethod()) {
+					exceptionTypes = mhi.reflectAs(Method.class, Lookup.IMPL_LOOKUP).getExceptionTypes();
+				} else if (mhi.isField()) {
+					/* Field reference kind has no checked exceptions. */
+				} else {
+					throw new InternalError("Unknown reference kind."); //$NON-NLS-1$
+				}
+				if (exceptionTypes != null) {
+					if (exceptionTypesList == null) {
+						if (relatedMHsList == null) {
+							return exceptionTypes;
+						} else {
+							exceptionTypesList = new ArrayList<>();
+						}
+					}
+					Collections.addAll(exceptionTypesList, exceptionTypes);
+				}
+			} else {
+				if (relatedMHsList == null) {
+					relatedMHsList = new ArrayList<>();
+				}
+				findRelatedMHs(mh, relatedMHsList);
+			}
+			if ((relatedMHsList == null) || relatedMHsList.isEmpty()) {
+				break;
+			} else {
+				mh = relatedMHsList.remove(relatedMHsList.size() - 1);
+			}
+		}
+		Class<?>[] exceptionTypesArray = null;
+		if (exceptionTypesList != null) {
+			exceptionTypesArray = exceptionTypesList.toArray(new Class<?>[exceptionTypesList.size()]);
+		}
+		return exceptionTypesArray;
+	}
+	
+	/**
+	 * A MethodHandle can be combination of multiple MethodHandles. This helper finds all the related
+	 * MethodHandles for a MethodHandle and appends the related MethodHandles to the relatedMHs list.
+	 * 
+	 * @param mh The MethodHandle to evaluate.
+	 * @param[out] relatedMHs A list of related MethodHandles.
+	 * 
+	 * @return void.
+	 * @throws InternalError for an unknown MethodHandle kind or an exception during a privileged
+	 * operation for ReceiverBoundHandle.
+	 */
+	private static void findRelatedMHs(MethodHandle mh, List<MethodHandle> relatedMHs) {
```

Comment:
> Is this method as in the RI?

No. It has similar functionality to [VarHandles.noCheckedExceptions](https://github.com/ibmruntimes/openj9-openjdk-jdk15/blob/dfa96ece90a70abcbf44415ca12537d7e531d810/src/java.base/share/classes/java/lang/invoke/VarHandles.java#L596-L633).

> This avoids needing to make the fields non-private

This method was derived from [SecurityFrameInjector.penetrateSecurityFrame](https://github.com/eclipse/openj9/blob/0ba7c58e810ebf0c730f1cbd9cc9191244f7bf2a/jcl/src/java.base/share/classes/java/lang/invoke/SecurityFrameInjector.java#L239-L319), which directly accesses the package-protected MH fields. There are discrepancies in OpenJ9 MHs where identical fields, such as `next`, are private and package-protected in some cases; example: `FilterArgumentsHandle.next` (private) and `FilterReturnHandle.next` (package-protected). `SecurityFrameInjector.penetrateSecurityFrame` is also in need of an update since it does not cover all of the OpenJ9 MHs. Do we ignore this discrepancy?

> The OO way ...

I like the OO approach (visitor pattern). I will be happy to take it. Shall I preserve the current implementation (commit) for reference and verifying behavior?

fyi - @DanHeidinga